core/src/main/antlr4/org/apache/druid/math/expr/antlr/Expr.g4

```diff
@@ -23,13 +23,20 @@ expr : 'null'                                         # null
      | expr ('<'|'<='|'>'|'>='|'=='|'!=') expr        # logicalOpExpr
      | expr ('&&'|'||') expr                          # logicalAndOrExpr
      | '(' expr ')'                                   # nestedExpr
+     | IDENTIFIER '(' lambda ',' fnArgs ')'           # applyFunctionExpr
      | IDENTIFIER '(' fnArgs? ')'                     # functionExpr
      | IDENTIFIER                                     # identifierExpr
      | DOUBLE                                         # doubleExpr
      | LONG                                           # longExpr
      | STRING                                         # string
+     | '[' DOUBLE  (','? DOUBLE)* ']'                 # doubleArray
+     | '[' LONG (','? LONG)* ']'                      # longArray
+     | '[' STRING (','? STRING)* ']'                  # stringArray
      ;
 
+lambda : (IDENTIFIER | '(' IDENTIFIER (','? IDENTIFIER)* ')') '->' expr
```

Comment:
Currently I don't have any use cases for no argument lambda's, but it would probably be friendlier to at least parse correctly and fail with invalid number of arguments error than the antlr stuff that shows up on stderr, will modify to allow `() ->`.

Reply:
I think this is still unresolved.
Also, I do believe that expression language should support no arg lambda because expressions are submitted by user and they can, for example, have an use case of converting all elements of array to some constant value. As a user, I would expect that to work.

Reply:
Oops! I think I added grammar for it, but forgot to do much more. Currently it does now parse into a `LambdaExpr`, but then for example
```
map(() -> 1, [1, 2, 3, 4, 5])
``` 
explodes in the form of:

```
java.lang.IllegalStateException: LambdaExpr has no or multiple arguments
```

because `map` calls `LambdaExpr.getIdentifier` which has a precondition that the `LambdaExpr` has 1 argument.

However, that was already there, I didn't do anything besides allow a no argument lambda to parse into a `LambdaExpr`, so amusingly
```
cartesian_map(() -> 1, [1, 2], [1, 2, 3])
```
does what you might expect, and makes a list of `[1, 1, 1, 1, 1, 1]`.

To me it maybe really only makes sense for `map` and maybe `cartesian_map` to support no argument lambda. Others like the filtering functions seem less useful, and i'm not sure what the fold functions would do, I'd assume they would still need to specify the accumulator identifier?

I'll fix up to explicitly handle no argument lambda for the maps, and explicitly guard to have the correct number of arguments for the cases where it doesn't make sense like filters and folds, and add tests for all cases.